File: 2020/day10.js
const fs = require('fs');

fs.readFile('input/day10', 'utf8', (_, data) => {
  const numbers = data.trim().split('\n').map(Number).sort((a, b) => a - b);
  const adapters = [0, ...numbers, numbers[numbers.length - 1] + 3];

  let oneDiffs = 0;
  let threeDiffs = 0;

  for (let i = 1; i < adapters.length; i++)
    if (adapters[i] - 1 === adapters[i - 1])
      oneDiffs ++;
    else if (adapters[i] - 3 === adapters[i - 1])
      threeDiffs ++;

  console.log('Part 1:', oneDiffs * threeDiffs);

  const countArrangementsFrom = (i, possibleArrangements = {}) => {
    if (i == adapters.length - 1) return 1;
    if (possibleArrangements[i]) return possibleArrangements[i];

    let ans = 0;

    for (let j = i + 1; j < adapters.length && adapters[j] - adapters[i] <= 3; j++)
      ans += countArrangementsFrom(j, possibleArrangements);

    possibleArrangements[i] = ans
    return ans;
  };

  console.log('Part 2:', countArrangementsFrom(0));
});

File: 2020/day10.lua
function main ()
  local adapters = {0}
  local one_diffs = 0
  local three_diffs = 0
  local possible_arrangements = {}

  for line in io.lines("input/day10") do
    table.insert(adapters, tonumber(line))
  end

  table.sort(adapters)
  table.insert(adapters, adapters[#adapters] + 3)

  for i, adapter in ipairs(adapters) do
    if adapter - 1 == adapters[i - 1] then one_diffs = one_diffs + 1 end
    if adapter - 3 == adapters[i - 1] then three_diffs = three_diffs + 1 end
  end

  print("Part 1:", one_diffs * three_diffs)
  print('Part 2:', count_arrangements_from(1, adapters, possible_arrangements))
end

function count_arrangements_from (i, adapters, possible_arrangements)
  if i == #adapters then return 1 end
  if possible_arrangements[i] then return possible_arrangements[i] end

  local ans = 0

  for j = i + 1, #adapters, 1 do
    if adapters[j] - adapters[i] <= 3 then
      ans = ans + count_arrangements_from(j, adapters, possible_arrangements)
    end
  end

  possible_arrangements[i] = ans
  return ans
end

main()
